src/adapter/src/catalog/storage.rs

```diff
@@ -838,6 +838,27 @@ impl<S: Append> Connection<S> {
         Ok(GlobalId::User(id))
     }
 
+    /// Get the next user and replica id without allocating them.
+    pub async fn get_next_ids(&mut self) -> Result<(GlobalId, ReplicaId), Error> {
```

Comment:
Bundling the fetching of these two IDs into a single API method corrupts the public API of the catalog with the specific needs of the `Controller`. Could you split this into two methods: `get_next_replica_id` and `get_next_user_global_id`?